tiny_dnn/core/framework/tensor.h

```diff
@@ -0,0 +1,516 @@
+            step = (to - from) / (host_data_->end() - host_data_->begin() - 1);
+        std::generate(host_data_->begin(),
+                      host_data_->end(),
+                      [&start, &step]() {
+                        U tmp = start;
+                        start += step;
+                        return tmp;
+                      });
+    }
+
+    /* @brief Element-wise addition
+     */
+    Tensor<U> add(const Tensor<U>& src) const {
+        if (this->shape() != src.shape()) {
+            throw nn_error("Tensor must have same shape");
+        }
+
+        Tensor<U> res(src.shape());
+
+        for_i(true, res.size(), [&](size_t i) {
+            res[i] = this->operator[](i) + src[i];
+        });
+
+        return std::move(res);
+    }
+
+    /* @brief Element-wise addition
+     */
+    Tensor<U> add(const float_t scalar) const {
+        Tensor<U> res(this->shape());
+
+        for_i(true, res.size(), [&](size_t i) {
+            res[i] = this->operator[](i) + scalar;
+        });
+
+        return std::move(res);
+    }
+
+    /* @brief Element-wise subtraction
+     */
+    Tensor<U> sub(const Tensor<U>& src) const {
+        if (this->shape() != src.shape()) {
+            throw nn_error("Tensor must have same shape");
+        }
+
+        Tensor<U> res(src.shape());
+
+        for_i(true, res.size(), [&](size_t i) {
+            res[i] = this->operator[](i) - src[i];
+        });
+
+        return std::move(res);
+    }
+
+    /* @brief Element-wise subtraction
+     */
+    Tensor<U> sub(const float_t scalar) const {
+        Tensor<U> res(this->shape());
+
+        for_i(true, res.size(), [&](size_t i) {
+            res[i] = this->operator[](i) - scalar;
+        });
+
+        return std::move(res);
+    }
+
+    /* @brief Element-wise multiplication
+     */
+    Tensor<U> mul(const Tensor<U>& src) const {
+        if (this->shape() != src.shape()) {
+            throw nn_error("Tensor must have same shape");
+        }
+
+        Tensor<U> res(src.shape());
+
+        for_i(true, res.size(), [&](size_t i) {
+            res[i] = this->operator[](i) * src[i];
+        });
+
+        return std::move(res);
+    }
+
+    /* @brief Element-wise multiplication
+     */
+    Tensor<U> mul(const float_t scalar) const {
+        Tensor<U> res(this->shape());
+
+        for_i(true, res.size(), [&](size_t i) {
+            res[i] = this->operator[](i) * scalar;
+        });
+
+        return std::move(res);
+    }
+
+    /* @brief Element-wise division
+     */
+    Tensor<U> div(const Tensor<U>& src) const {
+        if (this->shape() != src.shape()) {
+            throw nn_error("Tensor must have same shape");
+        }
+
+        Tensor<U> res(src.shape());
+
+        for_i(true, res.size(), [&](size_t i) {
+            const U tmp = src[i];
+            res[i] = tmp == U(0.0) ? std::numeric_limits<U>::quiet_NaN() :
+                this->operator[](i) / (tmp + std::numeric_limits<U>::min());
```

Comment:
> Looks like VS somewhere tries to copy unique_ptr:

maybe because the destructor is missing ? I don't have a MS environment to test :(

> Now you don't need addition

thx!